renter/rpc.go

```diff
@@ -206,3 +206,271 @@ func (s *Session) ScanSettings() (rhp.HostSettings, error) {
+		return rhp.Contract{}, types.Transaction{}, errors.New("host final revision signature is invalid")
+	} else if !s.hostKey.VerifyHash(vc.ContractSigHash(initialRevision), resp.InitialSignature) {
+		return rhp.Contract{}, types.Transaction{}, errors.New("host initial contract signature is invalid")
+	}
+
+	renewal.HostRollover = resp.HostRollover
+	renewal.FinalRevision.HostSignature = resp.FinalizationSignature
+	renewal.InitialRevision.HostSignature = resp.InitialSignature
+	// verify the host's renewal signature
+	renewalSigHash := vc.RenewalSigHash(*renewal)
+	if !s.hostKey.VerifyHash(renewalSigHash, resp.RenewalSignature) {
+		return rhp.Contract{}, types.Transaction{}, errors.New("host renewal signature is invalid")
+	}
+
+	// now that all other signatures are complete the renewal can be signed.
+	renewal.HostSignature = resp.RenewalSignature
+	renewal.RenterSignature = renterKey.SignHash(renewalSigHash)
+	// add the host's inputs to the transaction.
+	renewalTxn.SiacoinInputs = append(renewalTxn.SiacoinInputs, resp.Inputs...)
+	renewalTxn.SiacoinOutputs = append(renewalTxn.SiacoinOutputs, resp.Outputs...)
+	hostInputs := renewalTxn.SiacoinInputs[len(renterInputs):]
+
+	// sign any inputs we added and send the signatures to the host.
+	renterSigs := rhp.RPCRenewContractRenterSignatures{
+		SiacoinInputSignatures: make([][]types.Signature, len(renterInputs)),
+		RenewalSignature:       renewal.RenterSignature,
+	}
+	if len(renterInputs) > 0 {
+		toSign := make([]types.ElementID, len(renterInputs))
+		for i, sci := range renterInputs {
+			toSign[i] = sci.Parent.ID
+		}
+		if err := s.wallet.SignTransaction(vc, &renewalTxn, toSign); err != nil {
+			return rhp.Contract{}, types.Transaction{}, fmt.Errorf("failed to sign renewal transaction: %w", err)
+		}
+		for i := range renterInputs {
+			renterSigs.SiacoinInputSignatures[i] = renewalTxn.SiacoinInputs[i].Signatures
+		}
+	}
+	if err := rpc.WriteResponse(stream, &renterSigs); err != nil {
+		return rhp.Contract{}, types.Transaction{}, fmt.Errorf("failed to write renter signatures: %w", err)
+	}
+
+	// add the host signatures to the transaction
+	var hostSigs rhp.RPCContractSignatures
+	if err := rpc.ReadResponse(stream, &hostSigs); err != nil {
+		return rhp.Contract{}, types.Transaction{}, fmt.Errorf("failed to read host signatures: %w", err)
+	} else if len(hostSigs.SiacoinInputSignatures) != len(hostInputs) {
+		return rhp.Contract{}, types.Transaction{}, fmt.Errorf("host sent %v signatures, expected %v", len(hostSigs.SiacoinInputSignatures), len(hostInputs))
+	}
+	for i := range hostInputs {
+		hostInputs[i].Signatures = hostSigs.SiacoinInputSignatures[i]
+	}
+
+	// the transaction should now be valid
+	if err := vc.ValidateTransaction(renewalTxn); err != nil {
+		return rhp.Contract{}, types.Transaction{}, fmt.Errorf("renewal transaction is invalid: %w", err)
+	}
+
+	return rhp.Contract{
+		ID:       renewalTxn.FileContractID(0),
```

Comment:
I opted to just call `FileContractID` here instead of constructing the ID manually. We could also add something like `FileContractRenewalID` but that feels awkward. I think reusing `FileContractID` is Fine, Actually, as long as we document it in the docstring.